Refactor code to support ViewModels

northcott-j opened the pull request:
**Context**
I'm working on generating HTML outputs from the Facebook data with Bootstrapped tables and charts (see #19). The data for the HTML pages will be nearly identical to what is exported to Excel, and with the previous version of the code, I'd have to calculate everything twice - once for Excel and once for HTML. I decided to refactor the code to make the Model's used to generate a View to be their own object so they can be shared.

**Changes**
Factored everything related to creating a View of the data out into separate mixins and classes. There's now a single ViewModel per View (a View = excel sheet) in an Analyzeable.

There should be 0 functional or performance changes from this PR. This was just a refactor to make adding HTML outputs easier. I decided to separate everything into two PR's to avoid creating one massive one.

thnukid commented:
```bash
Finished in 0.00489 seconds (files took 1.07 seconds to load)
8 examples, 7 failures

Failed examples:

rspec ./spec/lib/facebook_data_analyzer/analyzeables/friends_spec.rb:27 # FacebookDataAnalyzer::Friends#count should count adding friends correctly
rspec ./spec/lib/facebook_data_analyzer/analyzeables/friends_spec.rb:11 # FacebookDataAnalyzer::Friends#analyze should return proper friends details
rspec ./spec/lib/facebook_data_analyzer/analyzeables/friends_spec.rb:19 # FacebookDataAnalyzer::Friends#total should return correct friends count
rspec ./spec/lib/facebook_data_analyzer/analyzeables/messages_spec.rb:40 # FacebookDataAnalyzer::Messages#group should group messages by conversation correctly
rspec ./spec/lib/facebook_data_analyzer/analyzeables/messages_spec.rb:25 # FacebookDataAnalyzer::Messages#count should count grouped messages correctly
rspec ./spec/lib/facebook_data_analyzer/analyzeables/messages_spec.rb:11 # FacebookDataAnalyzer::Messages#analyze should return proper @messages
rspec ./spec/lib/facebook_data_analyzer/analyzeables/contacts_spec.rb:11 # FacebookDataAnalyzer::Contacts#analyze should return proper @contacts details for first contact

Randomized with seed 573
```

Benchmark for `northcott-j:refactor-to-support-view-models`:
```
 bin/facebook_data_analyzer -b
       user     system      total        real
                                                                                                                                                              Analyzing Messages |Time: 00:00:00 | ======================================================================================================== | Time: 00:00:00
= Export facebook_analysis.xlsx
= Export facebook_analysis.html
  0.330000   0.030000   0.360000 (  0.357670)
```

Against Master:
```
in/facebook_data_analyzer -b
       user     system      total        real
                                                                                                                                                              Analyzing Messages |Time: 00:00:00 | ======================================================================================================== | Time: 00:00:00
= Export facebook_analysis.xlsx
= Export facebook_analysis.html
  0.320000   0.020000   0.340000 (  0.335635)
```

Lackoftactics commented:
@northcott-j could you resolve the conflict?

northcott-j commented:
@lackoftactics - will do! I planned on resolving them when I find some time to fix the tests

Lackoftactics commented:
@northcott-j any update?

northcott-j commented:
@Lackoftactics - sorry for the delay. It took much longer than expected to get around to picking this back up. I fixed the tests. As for perf differences, it takes me `32.65` seconds to run the current master code with my archive. It takes `33.16` with my current code MINUS the 3 new sheets I added. With those new sheets, it takes `37.91` seconds.

*Bug fix included*: Found a mistake with processing message HTML files where a message would be assigned to an incorrect sender if it followed an unexpected HTML block (e.g. images, stickers, files). This is fixed by skipping the last seen sender if a valid message isn't found before a new sender header is seen. This may drop a handful of messages, but I figured that would be better than falsely assigning them. After this change, I found that it changed my larger message counts (~18k+) by 5 or 6 messages, so not a huge difference.

The sheets added are:
* Percent Words In Common: Takes your unique words across all conversations and a friend's unique words across all conversations, and finds the percent intersection.
* Number of Corrections: When I make a typo or say something confusing, I usually add a `*` to a second message to clarify or fix. This counts the number of messages containing `*`. Not super accurate, but I was curious what the results would be.
* Money Talk: Counts the number of messages that contain `$`. The message parsing gsub will have to be updated to whitelist different symbols for other currencies. 

I added a fourth, but it's currently disabled as it takes several seconds to run.
* Decline of Unique Words: This creates a table meant to produce a graph that shows the average percent of unique words per chunk of messages per person over time. I limited it to you and 9 other friends and used bucket sizes of 5 messages. Example output (had to use Google Sheets because Excel couldn't graph it. Graph coincidentally also shows frequency of messages sent by a person shown by how big gaps are between points.): 
<img width="1009" alt="screen shot 2018-06-08 at 10 39 42 am" src="https://user-images.githubusercontent.com/9503623/41164078-4dc2f4d4-6b08-11e8-96a0-6b48b5635096.png">

Lackoftactics commented:
@northcott-j I merged first @marzann PR and you have some conflicts to solve. Great work btw. :)

Lackoftactics commented:
I love how you found those bugs, this looks like a great addition to this library.

Lackoftactics commented:
@northcott-j would you mind repairing tests after I merged your commit everything failed or do you prefer for me to revert it. I know that it works, results are just different now.

northcott-j commented:
@Lackoftactics - I will try to find time to fix it ASAP, but I'd revert it so master stays green
